sunscreen_runtime/src/linked.rs

```diff
@@ -0,0 +1,611 @@
+     * twos complement binary representation. This is because in the SDLP, the
+     * message polynomial is expanded into its twos complement binary and then used
+     * as an input to the proof. In order to link the SDLP and the ZKP program, we
+     * will be sharing this binary expansion between the two proof systems. This
+     * means that in the ZKP, we will need to convert the binary expanded message
+     * polynomial back into something meaningful for us. In this particular example
+     * (a constant polynomial message with bounds on the DC component only), we can
+     * use this helper function to reconstitute the transaction amount.
+     *
+     * ```
+     * fn from_twos_complement_field_element<F: FieldSpec, const N: usize>(
+     *     x: [ProgramNode<Field<F>>; N],
+     * ) -> ProgramNode<Field<F>> {
+     *     let mut x_recon = zkp_var!(0);
+     *
+     *     for (i, x_i) in x.iter().enumerate().take(N - 1) {
+     *         x_recon = x_recon + (zkp_var!(2i64.pow(i as u32)) * (*x_i));
+     *     }
+     *
+     *     x_recon = x_recon + zkp_var!(-(2i64.pow((N - 1) as u32))) * x[N - 1];
+     *
+     *     x_recon
+     * }
+     * ```
+     *
+     * With all of these pieces, we can use the `linked_proof` function to generate
+     * a proof that the encrypted transaction amount is less than or equal to the
+     * balance.
+     *
+     * ```
+     * let app = Compiler::new()
+     *     .zkp_backend::<BulletproofsBackend>()
+     *     .zkp_program(valid_transaction)
+     *     .compile()?;
+     *
+     * let valid_transaction_zkp = app.get_zkp_program(valid_transaction).unwrap();
+     *
+     * let transaction = 11999u64;
+     * let balance = 12200u64;
+     *
+     * let lattice_problem = test_seal_linear_relation::<SealQ128_1024, 1>(
+     *     transaction, 1024, 12289
+     * );
+     *
+     * // This means we only care about S[(0, 0)], which is `m` in the BFV encryption.
+     * let shared_indices = vec![(0, 0)];
+     *
+     * println!("Performing linked proof");
+     * let lp: LinkedProof = linked_proof(
+     *     &lattice_problem,
+     *     &shared_indices,
+     *     valid_transaction_zkp,
+     *     &[],                                 // Additional private inputs
+     *     &[BulletproofsField::from(balance)], // Public inputs
+     *     &[],                                 // Constant inputs
+     * );
+     * println!("Linked proof done");
+     * ```
+     *
+     * This will generate an proof of type `LinkedProof` that can be verified as follows:
+     *
+     * ```
+     * println!("Performing linked verify");
+     * let verified = linked_verify(
+     *     &lp,
+     *     valid_transaction_zkp,
+     *     vec![BulletproofsField::from(balance)], // Public inputs
+     *     vec![],                                 // Constant inputs
+     * );
+     * println!("Verified linked proof: {}", verified);
+     * ```
+     */
+    pub fn create<I>(
```

Comment:
We'll need to expose a function for the prover and verifier independently, since the verifier won't have the witness.

Reply:
But generally I think this is almost right. It should take a prover or verifier knowledge and link the proofs.